crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -737,6 +730,175 @@ impl Object {
         Ok(conn)
     }
 
+    /// Query the database for a `page` of objects, optionally `filter`-ed.
+    ///
+    /// The `checkpoint_viewed_at` parameter is an Option<u64> representing the
+    /// checkpoint_sequence_number at which this page was queried for, or None if the data was
+    /// requested at the latest checkpoint. Each entity returned in the connection will inherit this
+    /// checkpoint, so that when viewing that entity's state, it will be from the reference of this
+    /// checkpoint_viewed_at parameter.
+    pub(crate) async fn paginate_historical(
+        db: &Db,
+        page: Page<Cursor>,
+        filter: ObjectFilter,
+        checkpoint_viewed_at: Option<u64>,
+    ) -> Result<Connection<String, Object>, Error> {
+        Self::paginate_subtype_historical(db, page, filter, checkpoint_viewed_at, Ok).await
+    }
+
+    pub(crate) async fn paginate_subtype_historical<T: OutputType>(
+        db: &Db,
+        page: Page<Cursor>,
+        filter: ObjectFilter,
+        checkpoint_viewed_at: Option<u64>,
+        downcast: impl Fn(Object) -> Result<T, Error>,
+    ) -> Result<Connection<String, T>, Error> {
+        let mut conn: Connection<String, T> = Connection::new(false, false);
+
+        // Regardless of whether the cursor is the upper or lower bound, the cursor's
+        // checkpoint_sequence_number identifies the consistent upper bound when it was calculated
+        let checkpoint_viewed_at =
+            match validate_cursor_consistency(checkpoint_viewed_at, page.after(), page.before()) {
+                Ok(checkpoint_viewed_at) => checkpoint_viewed_at,
+                Err(e) => return Err(e),
+            };
+
+        let response = db
+            .execute_repeatable(move |conn| {
+                let (lhs, mut rhs) = Checkpoint::available_range(conn)?;
+
+                if let Some(checkpoint_viewed_at) = checkpoint_viewed_at {
+                    if checkpoint_viewed_at < lhs || rhs < checkpoint_viewed_at {
+                        return Ok::<_, diesel::result::Error>(None);
+                    }
+                    rhs = checkpoint_viewed_at;
+                }
+
+                let result = page.paginate_raw_query::<StoredHistoryObject, _>(
+                    conn,
+                    move |element| {
+                        element.map(|obj| {
+                            Cursor::new(HistoricalObjectCursor::new(obj.object_id.clone(), rhs))
+                        })
+                    },
```

Comment:
I dug into things a bit more and I now see why we need this (*siiiiiiiigh*) it's because we compare against the page's `after` and `before` cursors when we are paginating the results, right? That is quite unfortunate, and I don't see a neat way around that, other than updating the `cursor` function on `Target` to accept a checkpoint sequence number, always. 

We could also add another function to that trait, with a default implementation:

```rust
/// Results from the database that are pointed to by cursors.
pub(crate) trait Target<C: CursorType> {
    type Source: QuerySource;

    /// The cursor pointing at this target value.
    fn cursor(&self) -> C;

    /// The cursor pointing at this target value, assuming it was read at `checkpoint_viewed_at`.
    fn consistent_cursor(&self, _checkpoint_viewed_at: u64) -> C {
        self.cursor();
    }
}

```

This way we can update the call-site to call `consistent_cursor` but we don't have to go around updating all the implementations to include a parameter that they never use.

Reply:
Right. I like the suggestion. Also, harkening back to your comment on the very first PR, I suppose both the `paginate_query` and `paginate_raw_query` case will need to be updated to accept a `checkpoint_viewed_at`. Gah this keeps blowing up in my face lol